utopia-remix/app/util/github-branch-contents.server.ts

```diff
@@ -0,0 +1,329 @@
+import { ServerEnvironment } from '../env.server'
+import * as os from 'os'
+import type { FileType } from './files'
+import { fileTypeFromFileName } from './files'
+import { assertNever } from './assertNever'
+import AWS from 'aws-sdk'
+import type { OctokitClient } from './github'
+import type { GetBranchProjectContentsRequest } from '../routes/internal.projects.$id.github.branches.$owner.$repo.branch.$branch'
+
+export type AssetToUpload = {
+  path: string
+  data: Buffer
+}
+
+type UnzipResult = {
+  projectContents: ProjectContentTreeRoot
+  assetsToUpload: AssetToUpload[]
+}
+
+export type UnzipEntry = unzipper.Entry & {
+  vars: {
+    uncompressedSize: number // for some reason this field is not available in the type defs, but it's there
+  }
+}
+
+export type BranchResponse = {
+  branch: {
+    branchName: string
+    originCommit: string
+    content: ProjectContentTreeRoot
+  }
+}
+
+export function getBranchProjectContents(params: {
+  projectId: string
+  owner: string
+  repo: string
+  branch: string
+  uploadAssets: boolean
+  existingAssets: ExistingAsset[]
+}) {
+  return async function (client: OctokitClient): Promise<ApiSuccess<BranchResponse>> {
+    // 1. get the branch details
+    const response = await client.request('GET /repos/{owner}/{repo}/branches/{branch}', {
+      owner: params.owner,
+      repo: params.repo,
+      branch: params.branch,
+    })
+    const commit = response.data.commit.sha
+
+    // 2. get the zipball
+    const tarball = await client.request('GET /repos/{owner}/{repo}/zipball/{ref}', {
+      owner: params.owner,
+      repo: params.repo,
+      ref: commit,
+    })
+
+    // 3. write the zipball to disk in a temporary folder
+    const { archiveName, zipFilePath } = writeZipballToTempFile({
+      commit: commit,
+      owner: params.owner,
+      repo: params.repo,
+      data: tarball.data as ArrayBuffer,
+    })
+
+    // 4. unzip the archive and process its entries
+    const { assetsToUpload, projectContents } = await unzipGithubArchive({
+      archiveName: archiveName,
+      zipFilePath: zipFilePath,
+      existingAssets: params.existingAssets,
+    })
+
+    // 6. if there are any assets to upload, upload them
+    if (params.uploadAssets) {
+      await uploadAssets({
+        assets: assetsToUpload,
+        projectId: params.projectId,
+      })
+    }
+
+    return toApiSuccess({
+      branch: {
+        branchName: params.branch,
+        originCommit: commit,
+        content: projectContents,
+      },
+    })
+  }
+}
+
+function writeZipballToTempFile(params: {
+  commit: string
+  owner: string
+  repo: string
+  data: ArrayBuffer
+}): { archiveName: string; zipFilePath: string } {
+  const tempDir = os.tmpdir()
+
+  const shortSha = params.commit.slice(0, 7)
+  const archiveName = `${params.owner}-${params.repo}-${shortSha}`
+
+  const zipFilePath = urlJoin(tempDir, archiveName + '.zip')
+  fs.writeFileSync(zipFilePath, Buffer.from(params.data as ArrayBuffer))
```

Comment:
This should be async

Reply:
Surely not as it would return a path that would potentially be partially written to?

Reply:
@Rheeseyb I used `writeFile` in https://github.com/concrete-utopia/utopia/pull/5639/commits/3a09618900fb3c2bada985b9f1009a1dd083350e

and then @seanparsons I added a uuid suffix to the temp filename just to be safe from concurrent loads https://github.com/concrete-utopia/utopia/pull/5639/commits/f92af75a981cf6ecd0933b3ba76c71b48b3c0dec